### PR TITLE
Add a CLI tool to fetch objects from cache and print their text representation

### PR DESCRIPTION
**Pull request by bduffany**
This is useful for fetching objects not associated with an invocation, for example workflows that failed to start, or when replaying actions in dev.

I've had this sitting in a branch for a while and have found a few uses for it, and a few other ppl have used it as well. Figured it's worth checking in at this point

---

**Version bump**: Patch

### CODE REVIEW

File: tools/cas/cas.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	blobDigest   = flag.String("digest", "", "Digest of the blob to fetch, in HASH/SIZE format.")
+	blobType     = flag.String("type", "", "Type of proto to inspect: Action, ActionResult, Command, stdout, or stderr.")
+	instanceName = flag.String("remote_instance_name", "", "Remote instance name")
+	target       = flag.String("target", "", "Cache grpc target, such as grpcs://remote.buildbuddy.io")
+	apiKey       = flag.String("api_key", "", "API key to attach to the outgoing context")
+
+	invocationID = flag.String("invocation_id", "", "Invocation ID. This is required when fetching the result of a failed action. Otherwise, it's optional.")
+)
+
+func main() {
```
Comment (luluz66) [inline tools/cas/cas.go:32]: would be nice to have a comment showing example usage

Comment (bduffany) [reply tools/cas/cas.go:49]: Done

**Comment by tylerwilliams** [comment]
It'd be cool to find a sensible way to slot this into the bb cli tool. Maybe we can do that in a future cl.

File: tools/cas/cas.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+//     bazel run //tools/cas -- -target=grpcs://remote.buildbuddy.dev -digest=HASH/SIZE -type=stderr
+//
+// Show a failed action result proto (requires invocation ID):
+//     bazel run //tools/cas -- -target=grpcs://remote.buildbuddy.dev -digest=HASH/SIZE -type=ActionResult -invocation_id=IID
+func main() {
+	flag.Parse()
+	if *target == "" {
+		log.Fatalf("Missing --target")
+	}
+	parts := strings.Split(*blobDigest, "/")
```
Comment (vadimberezniker) [inline tools/cas/cas.go:54]: Use `digest.Parse`

File: tools/cas/cas.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	if err != nil {
+		log.Fatalf("Error parsing digest: %s", err)
+	}
+	ind := digest.NewResourceName(&repb.Digest{
+		Hash:      parts[0],
+		SizeBytes: int64(sizeBytes),
+	}, *instanceName)
+	conn, err := grpc_client.DialTarget(*target)
+	if err != nil {
+		log.Fatalf("Error dialing CAS target: %s", err.Error())
```
Comment (vadimberezniker) [inline tools/cas/cas.go:68]: just `err` (%s understands errors)

File: tools/cas/cas.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	}
+	bsClient := bspb.NewByteStreamClient(conn)
+	acClient := repb.NewActionCacheClient(conn)
+
+	ctx := context.Background()
+	if *apiKey != "" {
+		ctx = metadata.AppendToOutgoingContext(ctx, "x-buildbuddy-api-key", *apiKey)
```
Comment (vadimberezniker) [inline tools/cas/cas.go:75]: Would be nice to add a utility function for this. We've been doing this in a bunch of different tools.

**Review by vadimberezniker — APPROVED** [review]
Awesome tool to have!

**Review by luluz66 — APPROVED** [review]

File: tools/cas/cas.go (hunk 4 above)
Comment (bduffany) [reply tools/cas/cas.go:71]: I agree -- would prefer do this as a separate PR though, probably at a later date when I have more time to work on CLI tools :)

File: tools/cas/cas.go (hunk 3 above)
Comment (bduffany) [reply tools/cas/cas.go]: Done

File: tools/cas/cas.go (hunk 2 above)
Comment (bduffany) [reply tools/cas/cas.go]: Done

File: tools/cas/cas.go (hunk 4 above)
Comment (vadimberezniker) [reply tools/cas/cas.go:71]: :+1:

**Pull request merged**